### ISSUE TITLE
cube_build runs out of memory

### ISSUE THREAD

**Issue by stscieisenhamer** [STScI-JWST/jwst#592]
With Jeb's data set, which includes 24 exposures, the cube_build.CubeBuildStep gets a 'kill 9'. From watching system resources, it is apparently due to memory exhaustion.

Log:
```
2017-01-13 08:37:43,136 - stpipe.MRSCubeAssoc - INFO - MRSCubeAssoc instance created.
2017-01-13 08:37:43,138 - stpipe.MRSCubeAssoc.cube_build - INFO - CubeBuildStep instance created.
2017-01-13 08:37:43,191 - stpipe.MRSCubeAssoc - INFO - Step MRSCubeAssoc running with args ('./IKEvD_cube_asn.json',).
2017-01-13 08:37:43,236 - stpipe.MRSCubeAssoc.cube_build - INFO - Step cube_build running with args ('./IKEvD_cube_asn.json',).
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Starting IFU Cube Building Step
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Input interpolation: pointcloud
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Coordinate system to use: ra-dec
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Weighting method for point cloud: standard
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Power Weighting distance : 2.000000
2017-01-13 08:37:43,237 - stpipe.MRSCubeAssoc.cube_build - INFO - Output Base IKEvD_cubeassoc
2017-01-13 08:39:33,979 - stpipe.MRSCubeAssoc.cube_build - INFO - The desired cubes covers the MIRI Channels: ['1', '1', '1', '2', '2', '2', '3', '3', '3', '4', '4', '4']
2017-01-13 08:39:33,979 - stpipe.MRSCubeAssoc.cube_build - INFO - The desried cubes covers the MIRI subchannels: ['SHORT', 'MEDIUM', 'LONG', 'SHORT', 'MEDIUM', 'LONG', 'SHORT', 'MEDIUM', 'LONG', 'SHORT', 'MEDIUM', 'LONG']
2017-01-13 08:39:33,979 - stpipe.MRSCubeAssoc.cube_build - INFO - Building Cube IKEvD_cubeassoc_ch3-4-1-2-mediumlongshort_s3d.fits
2017-01-13 08:39:33,979 - stpipe.MRSCubeAssoc.cube_build - INFO - Number of bands in cube  12
2017-01-13 08:39:33,980 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 1,SHORT
2017-01-13 08:39:35,403 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.5.s
2017-01-13 08:39:36,672 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:38,008 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:39,333 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:39,333 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 1,MEDIUM
2017-01-13 08:39:40,606 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:41,975 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:43,382 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:44,689 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:44,689 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 1,LONG
2017-01-13 08:39:46,001 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:47,330 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:48,601 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:50,804 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:50,804 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 2,SHORT
2017-01-13 08:39:52,076 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:53,399 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:54,735 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:56,041 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:56,041 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 2,MEDIUM
2017-01-13 08:39:57,303 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:58,553 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:39:59,842 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:01,097 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:01,097 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 2,LONG
2017-01-13 08:40:02,386 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:03,670 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:04,941 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:06,244 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:06,244 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 3,SHORT
2017-01-13 08:40:07,420 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:08,528 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:09,675 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:10,808 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:10,808 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 3,MEDIUM
2017-01-13 08:40:11,898 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:12,974 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:14,924 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:16,039 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:16,039 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 3,LONG
2017-01-13 08:40:17,134 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:18,279 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:19,435 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:20,556 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:20,556 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 4,SHORT
2017-01-13 08:40:21,640 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:22,731 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:23,857 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:24,930 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:24,930 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 4,MEDIUM
2017-01-13 08:40:26,032 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:27,110 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:28,197 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:29,290 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:29,290 - stpipe.MRSCubeAssoc.cube_build - INFO - Working on data  from 4,LONG
2017-01-13 08:40:30,427 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:31,546 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:32,660 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:33,723 - stpipe.MRSCubeAssoc.cube_build - INFO - Time find foot print = 0.3.s
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube Geometry:
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - axis# Naxis  CRPIX    CRVAL      CDELT(arc sec)  MIN & Max (xi,eta arc sec)
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Axis 1    76  38.00   0.00003382   0.17200100  -6.41964050   6.41964074
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Axis 2    54  27.00  -0.04723833   0.17720418  -4.71989376   4.71989360
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Axis 3 27162   1.00   4.84696453   0.00086975   4.84696453  28.47111403
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 1
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 1
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 1
2017-01-13 08:40:33,729 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 2
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 2
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 2
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 3
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 3
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 3
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 4
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 4
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers channel: 4
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: SHORT
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: MEDIUM
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: LONG
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: SHORT
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: MEDIUM
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: LONG
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: SHORT
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: MEDIUM
2017-01-13 08:40:33,741 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: LONG
2017-01-13 08:40:33,742 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: SHORT
2017-01-13 08:40:33,742 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: MEDIUM
2017-01-13 08:40:33,742 - stpipe.MRSCubeAssoc.cube_build - INFO - Cube covers subchannel: LONG
2017-01-13 08:40:36,747 - stpipe.MRSCubeAssoc.cube_build - INFO - Region of interest 0.172001 0.177204 0.000870
Killed: 9
```

**Comment by stscieisenhamer** [comment]
@jemorrison fyi

**Comment by jemorrison** [comment]
The cube is size 76 X 54 X 27162
So I can see that it could run out of memory. I need to improve upon that.
I am working on making it faster and I will look at memory issues too. But
you might try running it so the wavelength plane is not so large. It looks
like the wavelength scale is 0.0008 microns - you can change it with the
option
--scalew = 0.001 (for example)

Jane

**Comment by jemorrison** [comment]
Cube build has a set of defaults to use for the scaling in the 3
dimensions. So that is what it originally used. These defaults were
determined by David Law and Ben Sargent but only for cubes made from the
same channels and band. The data that is causing problems is data for all
the channels and all the bands. We have not looked at what the best scaling
should be for data coming from different channels yet.

Jane

**Comment by jemorrison** [comment]
But even at your sampling that would only be ~850 MB, which, while big, should easily be handled by systems with 8-16GB or RAM….
-Jeb
—— 
Dr. J.I. Bailey, III (Jeb) / Sterrewacht Leiden, Universiteit Leiden
jebbailey.com <http://jebbailey.com/> / +1 (734) 389-5143 / +31 (0) 6 57 656 300 / skype:spacecolonyone <skype:spacecolonyone>

**Comment by hbushouse** [comment]
I'm wondering if the issue isn't the data volume of the cube, but rather the memory occupied by having 24 input exposures loaded simultaneously (assuming they're all loaded simultaneously).

**Comment by jemorrison** [comment]
Howard I think you are correct. The cube build is going to have problems
with the point cloud method if all the data is stored in memory when there
are many exposures. I have thought about how to change this and they way
that it is implemented it would not be too difficult to remove storing the
point cloud. Currently I form the point cloud and then loop over every
member to find which spaxels are with the ROI. I can re-organize it to map
to sky and find with spaxels the pixels falls within the RIO of spaxel
center but not store each point cloud element. I also removed a number of
values from the spaxel class that was taking up memory that where really
not needed.

Jane

**Comment by stscieisenhamer** [comment]
@jemorrison Methinks this was resolved by #688 and #689? At least for me, it is fixed.

**Comment by jemorrison** [comment]
This issue was resolved with #689. The point cloud information is no longer stored in memory. The detector pixels that fall with a region of interest of a spaxel are dynamically added to the spaxel.
